### PR TITLE
Ferminet: adding torch.func and changes to model input type

### PR DESCRIPTION
**Pull request by shaipranesh2**
## Description
Have utilized torch.func's hessian and jacobian calculation, which avoids the bigger nested loops making the model suitable for long calculations. Have changed the model's input type and the way the forward is called.

Fix #(issue)

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/tests/test_ferminet.py [hunk 1]
```diff
@@ -35,7 +35,7 @@ def test_forward():
     FH_molecule = [['F', [0.424, 0.424, 0.23]], ['H', [0.4, 0.5, 0.5]]]
     # Testing ionic initialization
     mol = FerminetModel(FH_molecule, spin=1, ion_charge=-1)
-    result = mol.model.forward(mol.molecule.x)
+    result = mol.model.forward(torch.from_numpy(mol.molecule.x))
```
Comment (shaipranesh2) [inline deepchem/models/tests/test_ferminet.py:38]: model will accept torch tensor in this new change thus making it possible for hessian and jacobian calculation without the previous nested loop

File: deepchem/models/tests/test_ferminet.py [hunk 2]
```diff
@@ -67,10 +67,10 @@ def test_FerminetMode_energy():
...
     mol = FerminetModel(H2_molecule, spin=0, ion_charge=0)
-    mol.train(nb_epoch=50)
-    mol.model.forward(mol.molecule.x)
+    mol.train(nb_epoch=5)
     energy = mol.model.calculate_electron_electron(
     ) - mol.model.calculate_electron_nuclear(
     ) + mol.model.nuclear_nuclear_potential + mol.model.calculate_kinetic_energy(
     )
-    assert energy <= torch.tensor(1.0)
+    mean_energy = torch.mean(energy)
```
Comment (shaipranesh2) [inline deepchem/models/tests/test_ferminet.py:75]: making changes to check for the average energy

File: deepchem/models/torch_models/ferminet.py [hunk 3]
```diff
@@ -107,13 +107,13 @@ def __init__(self,
...
 
-    def forward(self, input: np.ndarray) -> torch.Tensor:
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
         """
         forward function
 
         Parameters:
         -----------
-        input: np.ndarray
+        input: torch.Tensor
```
Comment (shaipranesh2) [inline deepchem/models/torch_models/ferminet.py:116]: input will be torch.tensor for jacobian and hessian calculation

File: deepchem/models/torch_models/ferminet.py [hunk 4]
```diff
@@ -190,7 +187,7 @@ def calculate_nuclear_nuclear(self,) -> torch.Tensor:
              self.nuclear_charge.unsqueeze(1)),
             posinf=0.0,
             neginf=0.0)
-        potential = torch.sum(potential) / 2
+        potential = (torch.sum(potential) / 2).unsqueeze(0)
```
Comment (shaipranesh2) [inline deepchem/models/torch_models/ferminet.py:190]: changing the format of of potential energy functions to calculate energies per batch, as hessians' and jacobians' energy too are per batch

**Review by rbharath — APPROVED** [review]
LGTM

**Pull request merged**